### PR TITLE
[REVIEW] Update CTK to CUDA 11.0

### PR DESCRIPTION
**Pull request by mike-wendt**
In addition add NVIDIA EULA and update about section to reflect the contents of this package.

- [x] Builds for Linux and Windows
- [x] CUDA 11 GA (May 2020) - https://anaconda.org/nvidia/cudatoolkit/files?version=11.0.194
- [x] CUDA 11 Update 1 (Aug 2020) - https://anaconda.org/nvidia/cudatoolkit/files?version=11.0.221
- [ ] Fix run constraints, see comment below

This follows initial work in #7 for the CUDA 11RC

### CODE REVIEW

**Comment by jjhelmus** [comment]
@mike-wendt Can you rebase now that #6 has been merged.  

Also adding `__cuda >=11.0` as a run requirement will prevent installation on incompatible system including CentOS 6.

**Comment by jjhelmus** [comment]
I'm getting errors similar to `FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpnydooixm/cuda-toolkit/lib64'` when I build this on linux?  Should the recipe be looking for files in `.../lib` rather than `../lib64`?

**Comment by mike-wendt** [comment]
> I'm getting errors similar to `FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmpnydooixm/cuda-toolkit/lib64'` when I build this on linux? Should the recipe be looking for files in `.../lib` rather than `../lib64`?

I think fixing the merge in my latest commit fixes this as I revereted to using `--toolkit` which I had successfully working before. That said I'm unable to build due to https://github.com/conda/conda-build/issues/4000

@jjhelmus @jakirkham do you have any suggestions. Right now for testing I've been removing it but I know it is needed in the end package that is released.

**Comment by mike-wendt** [comment]
I know the reason this breaks now. The changes made and merged from #6 do not work with the CUDA 11 installer on linux or windows. I have both environments up for testing now and will continue to work on this. I'm working on setting up a ppc64le env as well to make sure this works there. I have to set this aside for the next few hours but will pick it up later today.

**Comment by mike-wendt** [comment]
> [conda/conda-build#4000](https://github.com/conda/conda-build/issues/4000)
> 
> @jjhelmus @jakirkham do you have any suggestions. Right now for testing I've been removing it but I know it is needed in the end package that is released.

The above is still a blocker for me even as I fix the file not found issues. How can we move forward on this fast once the scripts are updated and ready to go?

**Comment by jjhelmus** [comment]
> The above is still a blocker for me even as I fix the file not found issues. How can we move forward on this fast once the scripts are updated and ready to go?

As I hinted at in the issue, running conda build with the `--no-test` flag should allow the build to complete.  The tests will not be run as part of the build but can be run after the fact using `conda build --test /path_to_packages/package.tar.bz2`

**Comment by mike-wendt** [comment]
Verified that the changes from #6 merge and 10.2 no longer work with the current installer for CUDA 11. The linux extract changes for amd64 work on ppc64le as well. Making updates to change to CUDA 11 Update 1 after pushing packages to the `nvidia` channel for CUDA 11.0. Will be ready for review in the morning.

**Comment by mike-wendt** [comment]
These versions do not have any run constraints:

CUDA 11 GA - https://anaconda.org/nvidia/cudatoolkit/files?version=11.0.194
CUDA 11 Update 1 - https://anaconda.org/nvidia/cudatoolkit/files?version=11.0.221

Adding `__cuda>=11.0` broke all of our solves in gpuCI, trying to use `__glibc>=2.17` also broke when it should solve. I understand the need for a constraint to ensure CentOS 6 users do not pick up this package, but there has to be a way to prevent them getting the package an not breaking existing solves. I'm open to suggestions, but at this point both of those options do not work for us currently.

Error message using `___glibc>=2.17`:
```
  - feature:/linux-64::__glibc==2.23=0
  - feature:|@/linux-64::__glibc==2.23=0
  - cudatoolkit=11.0 -> __glibc[version='>=2.17']

Your installed version is: 2.23
```

Error using `__cuda>=11.0`:
```
  - cudatoolkit=11.0 -> __cuda[version='>=11.0']

Your installed version is: not available
```

Both of these were from docker builds running on a CPU-only node, but building CUDA images. As I mentioned in a commit we need something that works for CPU-only environments as well given we build all of our conda packages on CPU-only nodes with CUDA images.

cc @kkraus14 @jakirkham

**Comment by jakirkham** [comment]
What does `conda info` say?

**Comment by kkraus14** [comment]
@jjhelmus would it be possible for you to review this again in the near future? We're trying to push out the RAPIDS 0.15 release and this update is needed to build CUDA 11 enabled conda packages in conda-forge for things like CuPy.

Is there anything we can do on our end to help reduce the maintenance burden on you?

**Comment by jjhelmus** [comment]
@kkraus14 @mike-wendt This looks good outside of the question on how to constrain the package.  I've been able to replicate the build on our machines for linux-64 and am trying a build on our linux-ppc64le machine as well.  

I need to do some more testing around the `__cuda` and `__glibc` virtual packages.  They are not behaving as I expect.  Would it be reasonable to have include `run_constained: _cuda >=11.0` here for the time being and if necessary hotfix this with something different if needed.  The existing `cudatoolkit` packages use constrained to partially enforce the driver requirement.

**Comment by jjhelmus** [comment]
Was able to confirm this builds fine on `linux-ppc64le` as well.  The proposed change to match the other `cudatoolkit` packages is:
``` diff
$ git diff 
diff --git a/meta.yaml b/meta.yaml
index 1f5361f..6ad4d6c 100644
--- a/meta.yaml
+++ b/meta.yaml
@@ -32,8 +32,8 @@ requirements:
     - tqdm
     # for run_exports
     - {{ compiler('cxx') }}
-  #run:
-  #  - __glibc >=2.17 # [linux]
+  run_constrained:
+    - _cuda >=11.0
```

I have linux packages built with this change that I plan on uploaded to `defaults` tonight/early tomorrow morning unless there are concerns.

**Comment by jjhelmus** [comment]
https://github.com/mike-wendt/conda-recipe-cudatoolkit/pull/4

**Comment by mike-wendt** [comment]
@jjhelmus this is not in master currently - how did the ppc64le 10.2 pkg get published and have the constraint on the web but the tarball and included `meta.yaml` not have it?

File: meta.yaml [hunk 1]
```diff
@@ -32,9 +32,26 @@ requirements:
     - tqdm
     # for run_exports
     - {{ compiler('cxx') }}
+  run_constrained:
+    - _cuda >=11.0
```
Comment (jakirkham) [inline meta.yaml:36]:
```suggestion
    - __cuda >=11.0
```

Comment (mike-wendt) [reply meta.yaml:36]: This breaks our builds `__cuda` as shown above

Comment (jakirkham) [reply meta.yaml:36]: `_cuda` doesn't do anything AFAIK. It has to have 2 `_`s.

Comment (mike-wendt) [reply meta.yaml:36]: If we can't build with this inside of a `nvidia/cuda` container then this is a large problem

**Comment by jakirkham** [comment]
So I thought we were already hotfixing `cudatoolkit` packages in PR ( https://github.com/AnacondaRecipes/repodata-hotfixes/pull/81 ). Does that already do what we want or should we be doing something different?

**Comment by jjhelmus** [comment]
> @jjhelmus this is not in master currently - how did the ppc64le 10.2 pkg get published and have the constraint on the web but the tarball and included `meta.yaml` not have it?

The `constraint` gets added via a patch when the packages are indexed.  Details are in AnacondaRecipes/repodata-hotfixes#81

**Comment by jjhelmus** [comment]
> So I thought we were already hotfixing `cudatoolkit` packages in PR ( [AnacondaRecipes/repodata-hotfixes#81](https://github.com/AnacondaRecipes/repodata-hotfixes/pull/81) ). Does that already do what we want or should we be doing something different?

We are but ideally the packages would have the constraint included rather than patched in.

**Comment by mike-wendt** [comment]
> So I thought we were already hotfixing `cudatoolkit` packages in PR ( [AnacondaRecipes/repodata-hotfixes#81](https://github.com/AnacondaRecipes/repodata-hotfixes/pull/81) ). Does that already do what we want or should we be doing something different?

This is my point on why is this change necessary here when it is obvious it is being added elsewhere. Now I know where.

File: meta.yaml (hunk 1 above)
Comment (jakirkham) [reply meta.yaml]: We can disable by setting `CONDA_OVERRIDE_CUDA=None`.

Comment (jjhelmus) [reply meta.yaml]: If needed you can override the detection `__cuda` version by setting the  `CONDA_OVERRIDE_CUDA` environment variable.

**Comment by jakirkham** [comment]
Ok, that sounds fine.

On the building point, could you please check whether this ( https://github.com/AnacondaRecipes/cudatoolkit-feedstock/pull/9#discussion_r476790125 ) works, Mike?

**Comment by mike-wendt** [comment]
@jakirkham I have to rebuild this package and then try to build images which is an hour or more of work. Given we're in the middle of a release I don't have that time to troubleshoot this at the moment. If you're both happy with this then I would say merge and publish.

I still have pkgs without the constraint so we won't be impacted but my suspicion is this is a larger issue. From my view an image that is `FROM nvidia/cuda` and has `miniconda` installed should work and not fail with this constraint. That being said it looks like I have a workaround so I'm good.

File: meta.yaml (hunk 1 above)
Comment (jjhelmus) [reply meta.yaml]: Thanks for catching this @jakirkham.  I suspect some of the odd behavior I was see is due to me using a single `_` cuda.

**Comment by jjhelmus** [comment]
With conda 4.8.4 I'm able to build packages from either of these two recipes if CONDA_OVERRIDE_CUDA=11.0 is set the the shell prior to calling conda build:

constrained:

``` yaml
package:
  name: test
  version: 1.0.0
requirements:
  run_constrained:
    - __cuda >=11.0
test:
  commands:
    - echo "Hi"
```

run:

``` yaml
package:
  name: test
  version: 2.0.0
requirements:
  run:
    - __cuda >=11.0
test:
  commands:
    - echo "Hi"
```

test-1.0.0 (the constraned version) is not install-able with conda 4.8.4 on system without the CUDA 11 driver but it can with earlier version of conda.  test-2.0.0 is not install-able without the CUDA 11 driver with both conda 4.8.4 or earlier versions.

**Comment by jakirkham** [comment]
No worries @mike-wendt. Just trying to make sure you have a path forward 🙂

**Comment by jjhelmus** [comment]
Merging.  linux-64 and linux-ppc64le packages should be available on `default` tonight.  win-64 will need to wait until the end of the week.

**Pull request merged**